src/System.Management.Automation/engine/InternalCommands.cs

```diff
@@ -424,24 +431,59 @@ private void InitParallelParameterSet()
                 _taskJob = new PSTaskJob(
                     Parallel.ToString(),
                     ThrottleLimit);
+
+                return;
             }
-            else
+
+            // Set up for synchronous processing and data streaming.
+            _taskCollection = new PSDataCollection<System.Management.Automation.PSTasks.PSTask>();
+            _taskDataStreamWriter = new PSTaskDataStreamWriter(this);
+            _taskPool = new PSTaskPool(ThrottleLimit);
+            _taskPool.PoolComplete += (sender, args) =>
             {
-                _taskDataStreamWriter = new PSTaskDataStreamWriter(this);
-                _taskPool = new PSTaskPool(ThrottleLimit);
-                _taskPool.PoolComplete += (sender, args) =>
-                {
-                    _taskDataStreamWriter.Close();
-                };
-                if (TimeoutSeconds != 0)
-                {
-                    _taskTimer = new Timer(
-                        (_) => _taskPool.StopAll(),
-                        null,
-                        TimeoutSeconds * 1000,
-                        Timeout.Infinite);
-                }
+                _taskDataStreamWriter.Close();
+            };
+
+            // Create timeout timer if requested.
+            if (TimeoutSeconds != 0)
+            {
+                _taskTimer = new Timer(
+                    callback: (_) => { _taskCollection.Complete(); _taskPool.StopAll(); },
+                    state: null,
+                    dueTime: TimeoutSeconds * 1000,
+                    period: Timeout.Infinite);
             }
+
+            // Task collection handler.
+            System.Threading.ThreadPool.QueueUserWorkItem(
+                (_) =>
+                {
+                    while (true)
+                    {
+                        _taskCollection.WaitHandle.WaitOne();
+
+                        var isOpen = _taskCollection.IsOpen;
+
+                        try
+                        {
+                            foreach (var task in _taskCollection.ReadAll())
+                            {
+                                _taskPool.Add(task);
+                            }
+                        }
+                        catch
+                        { 
+                            _taskCollection.Complete();
+                        }
+
+                        if (!isOpen)
```

Comment:
Why is it that:

- We record `waitOpen`'s value above,
- *Then* add tasks to the task pool
- *Then* check the first value, rather than checking `_taskPool.IsOpen` there?

What's the reason we want to capture the old state, do stuff and then query the old state?

Maybe worth documenting?

Reply:
It is in case a new task is added and collection closed while processing current collection tasks, and then that task is never transferred to the pool.  Loop is exited only when task collection is closed and all tasks added.